File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/media/MediaLibraryModels.kt
package io.ente.photos.platform.media

import java.nio.charset.StandardCharsets

enum class MediaAssetKind {
    IMAGE,
    VIDEO,
}

enum class MediaThumbnailFit {
    CONTAIN,
    COVER,
}

data class MediaThumbnailRequest(
    val assetId: String,
    val kind: MediaAssetKind,
    val widthPx: Int,
    val heightPx: Int,
    val fit: MediaThumbnailFit,
    val quality: Int,
    val allowNetworkAccess: Boolean,
) {
    init {
        require(assetId.isNotEmpty() && assetId.toByteArray(StandardCharsets.UTF_8).size <= 1024)
        require(widthPx in 1..2048)
        require(heightPx in 1..2048)
        require(widthPx.toLong() * heightPx <= 4_000_000)
        require(quality in 1..100)
    }
}

data class MediaThumbnail(
    val jpegBytes: ByteArray,
    val widthPx: Int,
    val heightPx: Int,
) {
    init {
        require(
            jpegBytes.size >= 4 &&
                jpegBytes[0] == 0xff.toByte() &&
                jpegBytes[1] == 0xd8.toByte() &&
                jpegBytes[jpegBytes.lastIndex - 1] == 0xff.toByte() &&
                jpegBytes[jpegBytes.lastIndex] == 0xd9.toByte(),
        )
        require(widthPx in 1..2048 && heightPx in 1..2048)
        require(widthPx.toLong() * heightPx <= 4_000_000)
    }
}

enum class MediaLibraryError {
    CANCELLED,
    PERMISSION_DENIED,
    ASSET_NOT_FOUND,
    RESOURCE_UNAVAILABLE,
    NETWORK_UNAVAILABLE,
    UNSUPPORTED_FORMAT,
    INVALID_REQUEST,
    BUSY,
    UNSUPPORTED_PLATFORM,
    PLATFORM_FAILURE,
}

enum class MediaAssetLookup {
    PRESENT,
    ABSENT,
    UNAVAILABLE,
}

fun missingAssetError(
    hasFullLibraryAccess: Boolean,
    lookup: MediaAssetLookup,
): MediaLibraryError =
    when {
        !hasFullLibraryAccess -> MediaLibraryError.PERMISSION_DENIED
        lookup == MediaAssetLookup.ABSENT -> MediaLibraryError.ASSET_NOT_FOUND
        else -> MediaLibraryError.RESOURCE_UNAVAILABLE
    }

fun mediaReadPermissionMeansFullAccess(
    apiLevel: Int,
    selectedMediaAccessDeclared: Boolean,
    readPermissionGranted: Boolean,
): Boolean =
    readPermissionGranted && (apiLevel < 34 || selectedMediaAccessDeclared)

sealed interface MediaThumbnailOutcome {
    data class Success(val thumbnail: MediaThumbnail) : MediaThumbnailOutcome

    data class Failure(val error: MediaLibraryError) : MediaThumbnailOutcome
}

File: mobile/apps/photos/plugins/ente_photos_platform/ios/Sources/MediaLibraryCore/MediaLibraryModels.swift
import Foundation

public enum MediaAssetKind: Equatable, Sendable {
    case image
    case video
}

public enum MediaThumbnailFit: Equatable, Sendable {
    case contain
    case cover
}

public struct MediaThumbnailRequest: Equatable, Sendable {
    public let assetID: String
    public let kind: MediaAssetKind
    public let widthPx: Int
    public let heightPx: Int
    public let fit: MediaThumbnailFit
    public let quality: Int
    public let allowNetworkAccess: Bool

    public init(
        assetID: String,
        kind: MediaAssetKind,
        widthPx: Int,
        heightPx: Int,
        fit: MediaThumbnailFit,
        quality: Int,
        allowNetworkAccess: Bool
    ) throws {
        guard !assetID.isEmpty, assetID.lengthOfBytes(using: .utf8) <= 1024,
            (1...2048).contains(widthPx),
            (1...2048).contains(heightPx),
            widthPx * heightPx <= 4_000_000,
            (1...100).contains(quality)
        else {
            throw MediaLibraryError.invalidRequest
        }
        self.assetID = assetID
        self.kind = kind
        self.widthPx = widthPx
        self.heightPx = heightPx
        self.fit = fit
        self.quality = quality
        self.allowNetworkAccess = allowNetworkAccess
    }
}

public struct MediaThumbnail: Equatable, Sendable {
    public let jpegData: Data
    public let widthPx: Int
    public let heightPx: Int

    public init(jpegData: Data, widthPx: Int, heightPx: Int) throws {
        guard jpegData.count >= 4,
            jpegData.starts(with: [0xff, 0xd8]),
            jpegData.suffix(2).elementsEqual([0xff, 0xd9]),
            (1...2048).contains(widthPx),
            (1...2048).contains(heightPx),
            widthPx * heightPx <= 4_000_000
        else {
            throw MediaLibraryError.platformFailure
        }
        self.jpegData = jpegData
        self.widthPx = widthPx
        self.heightPx = heightPx
    }
}

public enum MediaLibraryError: Error, Equatable, Sendable {
    case cancelled
    case permissionDenied
    case assetNotFound
    case resourceUnavailable
    case networkUnavailable
    case unsupportedFormat
    case invalidRequest
    case busy
    case unsupportedPlatform
    case platformFailure
}

public enum MediaThumbnailOutcome: Equatable, Sendable {
    case success(MediaThumbnail)
    case failure(MediaLibraryError)
}
